File: crates/defi-core/src/traits/liquid_staking.rs
use async_trait::async_trait;

use crate::error::Result;
use crate::types::*;

#[async_trait]
pub trait LiquidStaking: Send + Sync {
    fn name(&self) -> &str;
    async fn build_stake(&self, params: StakeParams) -> Result<DeFiTx>;
    async fn build_unstake(&self, params: UnstakeParams) -> Result<DeFiTx>;
    async fn get_info(&self) -> Result<StakingInfo>;
}

File: ts/packages/defi-core/src/traits/liquid_staking.ts
import type { StakeParams, UnstakeParams, StakingInfo, DeFiTx } from "../types.js";

export interface ILiquidStaking {
  name(): string;
  buildStake(params: StakeParams): Promise<DeFiTx>;
  buildUnstake(params: UnstakeParams): Promise<DeFiTx>;
  getInfo(): Promise<StakingInfo>;
}
